File: apps/iosApp/Sources/ViewModels/PlayerViewModel.swift
import Foundation
import SharedKit

enum PlayerUiState {
    case loading
    case error(String)
    case ready(PlaybackGrant)
}

@MainActor
final class PlayerViewModel: ObservableObject {
    @Published private(set) var uiState: PlayerUiState = .loading

    private let contentId: String
    private let playbackRepository: PlaybackRepository

    init(contentId: String, playbackRepository: PlaybackRepository, activeProfileStore: ActiveProfileStore) {
        self.contentId = contentId
        self.playbackRepository = playbackRepository
        guard let profileId = activeProfileStore.activeProfileId else {
            uiState = .error("No profile selected.")
            return
        }
        Task {
            do {
                let command = StartPlaybackCommand(contentId: contentId, profileId: profileId)
                let grant = try await ThrowingWrappersKt.startOrThrow(playbackRepository, command: command, userId: "")
                uiState = .ready(grant)
            } catch {
                uiState = .error(readableMessage(error))
            }
        }
    }

    func reportPosition(positionSeconds: Int64, durationSeconds: Int64?, completed: Bool = false) {
        guard case .ready(let grant) = uiState else { return }
        Task {
            let boxedDuration = durationSeconds.map { KotlinLong(value: $0) }
            let command = UpdatePositionCommand(positionSeconds: positionSeconds, durationSeconds: boxedDuration, completed: completed)
            _ = try? await ThrowingWrappersKt.updatePositionOrThrow(playbackRepository, sessionId: grant.session.id, userId: "", command: command)
        }
    }
}

File: apps/androidApp/src/main/kotlin/com/streaming/platform/android/ui/PlayerViewModel.kt
package com.streaming.platform.android.ui

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.streaming.platform.android.auth.ActiveProfileHolder
import com.streaming.platform.playback.PlaybackGrant
import com.streaming.platform.playback.PlaybackRepository
import com.streaming.platform.playback.StartPlaybackCommand
import com.streaming.platform.playback.UpdatePositionCommand
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch

sealed interface PlayerUiState {
    data object Loading : PlayerUiState
    data class Error(val message: String) : PlayerUiState
    data class Ready(val grant: PlaybackGrant) : PlayerUiState
}

class PlayerViewModel(
    private val contentId: String,
    private val playbackRepository: PlaybackRepository,
    activeProfileHolder: ActiveProfileHolder,
) : ViewModel() {
    private val _uiState = MutableStateFlow<PlayerUiState>(PlayerUiState.Loading)
    val uiState: StateFlow<PlayerUiState> = _uiState.asStateFlow()

    private val profileId = activeProfileHolder.activeProfileId.value

    init {
        val activeProfileId = profileId
        if (activeProfileId == null) {
            _uiState.value = PlayerUiState.Error("No profile selected.")
        } else {
            viewModelScope.launch {
                playbackRepository.start(StartPlaybackCommand(contentId, activeProfileId), userId = "").fold(
                    onSuccess = { grant -> _uiState.value = PlayerUiState.Ready(grant) },
                    onFailure = { _uiState.value = PlayerUiState.Error(it.readableMessage()) },
                )
            }
        }
    }

    fun reportPosition(positionSeconds: Long, durationSeconds: Long?, completed: Boolean = false) {
        val state = _uiState.value as? PlayerUiState.Ready ?: return
        viewModelScope.launch {
            playbackRepository.updatePosition(
                state.grant.session.id,
                userId = "",
                UpdatePositionCommand(positionSeconds, durationSeconds, completed),
            )
        }
    }
}
